File: Data Structures & Algorithms/search-2d-matrix/submission-0.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty() || matrix[0].empty()) return false; // Check for empty matrix

    int rows = matrix.size();
    int cols = matrix[0].size();
    int left = 0, right = rows * cols - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2; // Calculate the mid index
        int midValue = matrix[mid / cols][mid % cols]; // Convert mid index to 2D coordinates

        if (midValue == target) {
            return true; // Target found
        } else if (midValue < target) {
            left = mid + 1; // Search in the right half
        } else {
            right = mid - 1; // Search in the left half
        }
    }
    return false;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0;
        int r = nums.length - 1;
        while (l <= r) {
            int mid = (l + r) / 2;

            if (nums[mid] == target) {
                return mid;
            }

            //if left half is sorted
            if (nums[l] <= nums[mid]) {
                //if its target is within range of left half
                if (target >= nums[l] && target < nums[mid]) {
                    //terminate right half
                    r = mid - 1;
                } else {
                    //else target is in right half
                    l = mid + 1;
                }
            } else {
                //same logic but in reverse
                if (target > nums[mid] && target <= nums[r]) {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }
        }

        return -1;
    }
}
